posts/2023-02-10-caching-in-github-actions.md

```diff
@@ -0,0 +1,122 @@
+
+It has the K8s API as a dependency which is quite large so we expect to save some time by cashing the Go mod download. By also caching the Go build cache it should be possible to speed up the workflow even more. 
+
+We will first create the workflow and run it without any caching.
+
+```yaml
+name: ci
+
+on: push
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          repository: "alexellis/registry-creds"
+      - name: Setup Golang
+        uses: actions/setup-go@v3
+        with:
+          go-version: ~1.19
+      - name: Build
+        run: |
+          CGO_ENABLED=0 GO111MODULE=on \
+          go build -ldflags "-s -w -X main.Release=dev -X main.SHA=dev" -o controller
+```
+
+The [checkout action](https://github.com/actions/checkout) is used to check out the registry-creds repo.The next step sets up Golang using the [setup-go action](https://github.com/actions/setup-go) and as a last step we run `go build`.
+
+When triggering this workflow we see that each run takes around `1m25s`.
+
+Modify the workflow and add an additional step to configure the caches using the [cache action](https://github.com/actions/cache):
+
+```yaml
+steps:
+  - name: Setup Golang
+    uses: actions/setup-go@v3
+    with:
+      go-version: ~1.19
+  - name: Setup Golang caches
+    uses: actions/cache@v3
+    with:
+      path: |
+        ~/.cache/go-build
+        ~/go/pkg/mod
+      key: ${{ runner.os }}-golang-${{ hashFiles('**/go.sum') }}
+      restore-keys: |
+        ${{ runner.os }}-golang-
+```
+
+The first time this workflow is run the cache is not populated so we see a similar run time as without any cache of around `1m25s`. Running the workflow again we can see that it now completes in just `24s`.
+
+The V3 edition of the [setup-go](https://github.com/actions/setup-go) action has support for caching built-in. Under the hood it also uses the [actions/cache](https://github.com/actions/cache) with a similar configuration as in the example above.
+
+The advantage of using the built-in functionality is that it requires less configuration settings. Caching can be enabled by adding a single line to the workflow configuration:
+
+```diff
+name: ci
+
+on: push
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          repository: "alexellis/registry-creds"
+      - name: Setup Golang
+        uses: actions/setup-go@v3
+        with:
+          go-version: ~1.19
++         cache: true
+      - name: Build
+        run: |
+          CGO_ENABLED=0 GO111MODULE=on \
+          go build -ldflags "-s -w -X main.Release=dev -X main.SHA=dev" -o controller
+```
+
+Triggering the workflow with the build-in cache yields similar time gains is with the manual cache configuration.
+
+## Conclusion
```

Comment:
I like the article a lot.

It finishes very suddenly though.

Think about additional links in the conclusion like the Go action on GitHub, a link to my eBook perhaps which has a good example for Go builds with Actions, a link to the generic caching action for GitHub, a word about how long the cache lasts?

Reply:
Mention other types of caching i.e. Docker layer caching, and link to that too.

Mention that it can use GitHub's cache but can also use the registry itself by publishing extra layers to the final image.

https://docs.docker.com/build/cache/backends/